### PR TITLE
Implement Response Streaming

### PR DESCRIPTION
**Pull request by carlodek**
## Motivation and Context (Why the change? What's the scenario?)

Add option to stream Ask result tokens without waiting for the full answer to be ready.

## High level description (Approach, Design)

- New `stream` boolean option for the `Ask` API, false by default. When true, answer tokens are streamed as soon as they are generated by LLMs.
- New `MemoryAnswer.StreamState` enum property: `Error`, `Reset`, `Append`, `Last`.
- If moderation is enabled, the content is validated at the end. In case of moderation failure, the service returns an answer with `StreamState` = `Reset` and the new content to show to the end user.
- Streaming uses SSE message format. 
- By default, SSE streams end with a `[DONE]` token. This can be disabled via KM settings.
- SSE payload is optimized, returning `RelevantSources` only in the first SSE message.

### CODE REVIEW

**Comment by carlodek** [comment]
@microsoft-github-policy-service agree

________________________________
Da: microsoft-github-policy-service[bot] ***@***.***>
Inviato: giovedì 1 agosto 2024 16:07
A: microsoft/kernel-memory ***@***.***>
Cc: carlodek ***@***.***>; Mention ***@***.***>
Oggetto: Re: [microsoft/kernel-memory] added ask_stream endpoint (PR #726)


@carlodek<https://github.com/carlodek> please read the following Contributor License Agreement(CLA). If you agree with the CLA, please reply with the following information.

@microsoft-github-policy-service agree [company="{your company}"]


Options:

  *   (default - no company specified) I have sole ownership of intellectual property rights to my Submissions and I am not making Submissions in the course of work for my employer.

@microsoft-github-policy-service agree


  *   (when company given) I am making Submissions in the course of work for my employer (or my employer has intellectual property rights in my Submissions by contract or applicable law). I have permission from my employer to make Submissions and enter into this Agreement on behalf of my employer. By signing below, the defined term “You” includes me and my employer.

@microsoft-github-policy-service agree company="Microsoft"


Contributor License Agreement
Contribution License Agreement

This Contribution License Agreement (“Agreement”) is agreed to by the party signing below (“You”),
and conveys certain license rights to Microsoft Corporation and its affiliates (“Microsoft”) for Your
contributions to Microsoft open source projects. This Agreement is effective as of the latest signature
date below.

  1.  Definitions.
“Code” means the computer software code, whether in human-readable or machine-executable form,
that is delivered by You to Microsoft under this Agreement.
“Project” means any of the projects owned or managed by Microsoft and offered under a license
approved by the Open Source Initiative (www.opensource.org<http://www.opensource.org>).
“Submit” is the act of uploading, submitting, transmitting, or distributing code or other content to any
Project, including but not limited to communication on electronic mailing lists, source code control
systems, and issue tracking systems that are managed by, or on behalf of, the Project for the purpose of
discussing and improving that Project, but excluding communication that is conspicuously marked or
otherwise designated in writing by You as “Not a Submission.”
“Submission” means the Code and any other copyrightable material Submitted by You, including any
associated comments and documentation.
  2.  Your Submission. You must agree to the terms of this Agreement before making a Submission to any
Project. This Agreement covers any and all Submissions that You, now or in the future (except as
described in Section 4 below), Submit to any Project.
  3.  Originality of Work. You represent that each of Your Submissions is entirely Your original work.
Should You wish to Submit materials that are not Your original work, You may Submit them separately
to the Project if You (a) retain all copyright and license information that was in the materials as You
received them, (b) in the description accompanying Your Submission, include the phrase “Submission
containing materials of a third party:” followed by the names of the third party and any licenses or other
restrictions of which You are aware, and (c) follow any other instructions in the Project’s written
guidelines concerning Submissions.
  4.  Your Employer. References to “employer” in this Agreement include Your employer or anyone else
for whom You are acting in making Your Submission, e.g. as a contractor, vendor, or agent. If Your
Submission is made in the course of Your work for an employer or Your employer has intellectual
property rights in Your Submission by contract or applicable law, You must secure permission from Your
employer to make the Submission before signing this Agreement. In that case, the term “You” in this
Agreement will refer to You and the employer collectively. If You change employers in the future and
desire to Submit additional Submissions for the new employer, then You agree to sign a new Agreement
and secure permission from the new employer before Submitting those Submissions.
  5.  Licenses.

  *   Copyright License. You grant Microsoft, and those who receive the Submission directly or
indirectly from Microsoft, a perpetual, worldwide, non-exclusive, royalty-free, irrevocable license in the
Submission to reproduce, prepare derivative works of, publicly display, publicly perform, and distribute
the Submission and such derivative works, and to sublicense any or all of the foregoing rights to third
parties.
  *   Patent License. You grant Microsoft, and those who receive the Submission directly or
indirectly from Microsoft, a perpetual, worldwide, non-exclusive, royalty-free, irrevocable license under
Your patent claims that are necessarily infringed by the Submission or the combination of the
Submission with the Project to which it was Submitted to make, have made, use, offer to sell, sell and
import or otherwise dispose of the Submission alone or with the Project.
  *   Other Rights Reserved. Each party reserves all rights not expressly granted in this Agreement.
No additional licenses or rights whatsoever (including, without limitation, any implied licenses) are
granted by implication, exhaustion, estoppel or otherwise.

  1.  Representations and Warranties. You represent that You are legally entitled to grant the above
licenses. You represent that each of Your Submissions is entirely Your original work (except as You may
have disclosed under Section 3). You represent that You have secured permission from Your employer to
make the Submission in cases where Your Submission is made in the course of Your work for Your
employer or Your employer has intellectual property rights in Your Submission by contract or applicable
law. If You are signing this Agreement on behalf of Your employer, You represent and warrant that You
have the necessary authority to bind the listed employer to the obligations contained in this Agreement.
You are not expected to provide support for Your Submission, unless You choose to do so. UNLESS
REQUIRED BY APPLICABLE LAW OR AGREED TO IN WRITING, AND EXCEPT FOR THE WARRANTIES
EXPRESSLY STATED IN SECTIONS 3, 4, AND 6, THE SUBMISSION PROVIDED UNDER THIS AGREEMENT IS
PROVIDED WITHOUT WARRANTY OF ANY KIND, INCLUDING, BUT NOT LIMITED TO, ANY WARRANTY OF
NONINFRINGEMENT, MERCHANTABILITY, OR FITNESS FOR A PARTICULAR PURPOSE.
  2.  Notice to Microsoft. You agree to notify Microsoft in writing of any facts or circumstances of which
You later become aware that would make Your representations in this Agreement inaccurate in any
respect.
  3.  Information about Submissions. You agree that contributions to Projects and information about
contributions may be maintained indefinitely and disclosed publicly, including Your name and other
information that You submit with Your Submission.
  4.  Governing Law/Jurisdiction. This Agreement is governed by the laws of the State of Washington, and
the parties consent to exclusive jurisdiction and venue in the federal courts sitting in King County,
Washington, unless no federal subject matter jurisdiction exists, in which case the parties consent to
exclusive jurisdiction and venue in the Superior Court of King County, Washington. The parties waive all
defenses of lack of personal jurisdiction and forum non-conveniens.
  5.  Entire Agreement/Assignment. This Agreement is the entire agreement between the parties, and
supersedes any and all prior agreements, understandings or communications, written or oral, between
the parties relating to the subject matter hereof. This Agreement may be assigned by Microsoft.

**Comment by dluc** [comment]
Update: for this feature to be merged, there's a couple of things to do:
* Check this similar PR https://github.com/microsoft/kernel-memory/pull/400 and decide which approach to take
* Support content moderation. The stream of tokens needs to be validated while streamed, on a configurable frequence. If at any point the text moderation fails, the stream needs to be reset, e.g. sending a special token or similar.

**Comment by nurkmez2** [comment]
I have been waiting for a long time if you could add Ask Stream Endpoint in the next version, it would be great

**Comment by nurkmez2** [comment]
Hi @carlodek 
It would be great if you could complete the feature of ask_stream endpoint.
Thanks and Regards

**Comment by carlodek** [comment]
Hello, I've noticed that your work works like a charm with Ollama too! Still with Ollama Stream call is working with ask_stream endpoint. @dluc let me know if I have to do something more

File: .dockerignore [hunk 1]
```diff
@@ -16,4 +16,4 @@ infra
 
 # To make sure that the local appsettings files are not copied when building the image locally.
 # These files are not in GitHub thanks to .gitignore
-**/*appsettings.*.json
+
```
Comment (dluc) [inline .dockerignore:19]:
```suggestion
**/*appsettings.*.json
```

Please revert this change. The folder can contain files like appsettings.Development.json and appsettings.Production.json that should not included in the docker image. The files can be imported using Docker volumes, see also the main README for an example.

File: .gitignore [hunk 2]
```diff
@@ -504,3 +504,4 @@ swa-cli.config.json
 
 *.orig
 .azure
+service/Service/AzureOpenAiDevelopment.json
```
Comment (dluc) [inline .gitignore:507]: I don't recognize this file, is it created by some tool?

File: Directory.Packages.props [hunk 3]
```diff
@@ -29,6 +29,8 @@
     <PackageVersion Include="Microsoft.ML.OnnxRuntimeGenAI.DirectML" Version="0.5.2" />
     <PackageVersion Include="MongoDB.Driver.GridFS" Version="2.30.0" />
     <PackageVersion Include="Moq" Version="4.20.72" />
+    <PackageVersion Include="Microsoft.Extensions.Logging.TraceSource" Version="8.0.0" />
+    <PackageVersion Include="Microsoft.KernelMemory.MemoryDb.Qdrant" Version="0.70.240803.1" />
```
Comment (dluc) [inline Directory.Packages.props:33]: where is this package used?

File: Dockerfile [hunk 4]
```diff
@@ -11,7 +11,7 @@ ARG RUN_IMAGE_TAG="8.0-alpine"
 
 FROM mcr.microsoft.com/dotnet/sdk:$BUILD_IMAGE_TAG AS build
 
-ARG BUILD_CONFIGURATION=Release
+ARG BUILD_CONFIGURATION=Debug
```
Comment (dluc) [inline Dockerfile:14]: why this change?

File: Dockerfile [hunk 5]
```diff
@@ -49,8 +49,7 @@ LABEL org.opencontainers.image.authors="Devis Lucato, https://github.com/dluc"
 USER $USER
 
 # Used by .NET and KM to load appsettings.Production.json
-ENV ASPNETCORE_ENVIRONMENT=Production
-ENV ASPNETCORE_URLS=http://+:9001
```
Comment (dluc) [inline Dockerfile:53]: this is required, why remove it?

File: service/Core/Core.csproj [hunk 6]
```diff
@@ -14,7 +14,7 @@
     </ItemGroup>
 
     <ItemGroup>
-        <PackageReference Include="ClosedXML" /> <!-- missing strong name (CS8002) -->
```
Comment (dluc) [inline service/Core/Core.csproj:17]: please undo this change

File: service/Service/Program.cs [hunk 7]
```diff
@@ -119,7 +119,7 @@ public static void Main(string[] args)
             });
 
         // CORS
-        bool enableCORS = false;
+        bool enableCORS = true;
```
Comment (dluc) [inline service/Service/Program.cs:122]: why this change?

**Review by dluc — CHANGES_REQUESTED** [review]
There are several changes out of scope and unnecessary ones, plus some affecting the solution security. First thing I would ask is to undo all these changes, including code style ones, limiting to the bare minimum. Thanks!

File: Dockerfile [hunk 8]
```diff
@@ -49,8 +49,7 @@ LABEL org.opencontainers.image.authors="Devis Lucato, https://github.com/dluc"
 USER $USER
 
 # Used by .NET and KM to load appsettings.Production.json
-ENV ASPNETCORE_ENVIRONMENT=Production
```
Comment (dluc) [inline Dockerfile:52]: KM docker image runs in Production mode

File: service/Service/Program.cs [hunk 9]
```diff
@@ -128,10 +128,11 @@ public static void Main(string[] args)
...
                 {
                     policy
+                        //.WithOrigins("http://127.0.0.1:5500") these three are to test in local for js and html, supposing you're opening 5500 port
+                        //.AllowAnyHeader()
+                        //.AllowAnyMethod()
                         .WithMethods("HEAD", "GET", "POST", "PUT", "DELETE")
                         .WithExposedHeaders("Content-Type", "Content-Length", "Last-Modified");
-                    // .AllowAnyOrigin()
-                    // .WithOrigins(...)
                     // .AllowAnyHeader()
```
Comment (dluc) [inline service/Service/Program.cs:136]: please undo

File: .gitignore (hunk 2 above)
Comment (dluc) [inline .gitignore:507]:
```suggestion
```

File: Directory.Packages.props (hunk 3 above)
Comment (dluc) [inline Directory.Packages.props:33]:
```suggestion
```

File: Dockerfile (hunk 4 above)
Comment (dluc) [inline Dockerfile:14]:
```suggestion
ARG BUILD_CONFIGURATION=Release
```

File: Dockerfile [hunk 10]
```diff
@@ -49,8 +49,7 @@ LABEL org.opencontainers.image.authors="Devis Lucato, https://github.com/dluc"
...
-ENV ASPNETCORE_URLS=http://+:9001
+ENV ASPNETCORE_ENVIRONMENT=Development
```
Comment (dluc) [inline Dockerfile:52]:
```suggestion
ENV ASPNETCORE_ENVIRONMENT=Production
```

File: Dockerfile [hunk 11]
```diff
@@ -50,7 +50,6 @@ USER $USER
 
 # Used by .NET and KM to load appsettings.Production.json
 ENV ASPNETCORE_ENVIRONMENT=Production
-ENV ASPNETCORE_URLS=http://+:9001
```
Comment (dluc) [inline Dockerfile:53]:
```suggestion
ENV ASPNETCORE_URLS=http://+:9001
```

File: Dockerfile [hunk 12]
```diff
@@ -50,7 +50,6 @@ USER $USER
 
 # Used by .NET and KM to load appsettings.Production.json
 ENV ASPNETCORE_ENVIRONMENT=Production
```
Comment (dluc) [inline Dockerfile:52]:
```suggestion
ENV ASPNETCORE_ENVIRONMENT=Production
ENV ASPNETCORE_URLS=http://+:9001
```
```

File: service/Core/Core.csproj [hunk 13]
```diff
@@ -14,7 +14,7 @@
...
+        <PackageReference Include="ClosedXML" />
```
Comment (dluc) [inline service/Core/Core.csproj:17]:
```suggestion
        <PackageReference Include="ClosedXML" /> <!-- missing strong name (CS8002) -->
```

File: service/Core/Search/SearchClient.cs [hunk 14]
```diff
@@ -212,7 +333,7 @@ private SearchClientResult ProcessMemoryRecord(
         string fileDownloadUrl = record.GetWebPageUrl(index);
 
         // Name of the file to show to the LLM, avoiding "content.url"
-        string fileNameForLLM = (fileName == "content.url" ? fileDownloadUrl : fileName);
+        string fileNameForLlm = (fileName == "content.url" ? fileDownloadUrl : fileName);
```
Comment (dluc) [inline service/Core/Search/SearchClient.cs:336]:
```suggestion
        string fileNameForLLM = (fileName == "content.url" ? fileDownloadUrl : fileName);
```

File: service/Core/Search/SearchClient.cs [hunk 15]
```diff
@@ -226,7 +347,7 @@ private SearchClientResult ProcessMemoryRecord(
             string fact = PromptUtils.RenderFactTemplate(
                 template: factTemplate!,
                 factContent: partitionText,
-                source: fileNameForLLM,
+                source: fileNameForLlm,
```
Comment (dluc) [inline service/Core/Search/SearchClient.cs:350]:
```suggestion
                source: fileNameForLLM,
```

File: service/Service/Program.cs [hunk 16]
```diff
@@ -128,10 +128,11 @@ public static void Main(string[] args)
                 options.AddPolicy(name: CORSPolicyName, policy =>
                 {
                     policy
+                        //.WithOrigins("http://127.0.0.1:5500") these three are to test in local for js and html, supposing you're opening 5500 port
+                        //.AllowAnyHeader()
+                        //.AllowAnyMethod()
```
Comment (dluc) [inline service/Service/Program.cs:133]:
```suggestion
```

File: service/Service/Program.cs (hunk 7 above)
Comment (dluc) [inline service/Service/Program.cs:122]:
```suggestion
        bool enableCORS = false;
```

File: service/Service/Program.cs [hunk 17]
```diff
@@ -130,8 +130,6 @@ public static void Main(string[] args)
                     policy
                         .WithMethods("HEAD", "GET", "POST", "PUT", "DELETE")
                         .WithExposedHeaders("Content-Type", "Content-Length", "Last-Modified");
```
Comment (dluc) [inline service/Service/Program.cs:132]:
```suggestion
                        .WithExposedHeaders("Content-Type", "Content-Length", "Last-Modified");
                    // .AllowAnyOrigin()
                    // .WithOrigins(...)
```

File: service/Core/Search/SearchClient.cs [hunk 18]
```diff
@@ -202,7 +323,7 @@ private SearchClientResult ProcessMemoryRecord(
         // Identify the file in case there are multiple files
         string fileId = record.GetFileId(this._log);
 
-        // Note: this is not a URL and perhaps could be dropped. For now it acts as a unique identifier. See also SourceUrl.
+        // Note: this is not a URL and perhaps could be dropped. For now, it acts as a unique identifier. See also SourceUrl.
```
Comment (dluc) [inline service/Core/Search/SearchClient.cs:326]:
```suggestion
        // Note: this is not a URL and perhaps could be dropped. For now it acts as a unique identifier. See also SourceUrl.
```

File: service/Service/Service.csproj [hunk 19]
```diff
@@ -53,4 +53,4 @@
         </Content>
     </ItemGroup>
 
-</Project>
\ No newline at end of file
+</Project>
```
Comment (dluc) [inline service/Service/Service.csproj:56]:
```suggestion
</Project>
```

File: service/Service.AspNetCore/WebAPIEndpoints.cs [hunk 20]
```diff
@@ -246,6 +247,59 @@ async Task<IResult> (
...
+                    HttpContext httpContext,
+                    MemoryQuery query,
+                    IKernelMemory service,
+                    ILogger<KernelMemoryWebAPI> log,
+                    IContextProvider contextProvider,
+                    CancellationToken cancellationToken) =>
+                {
+                    // Allow internal classes to access custom arguments via IContextProvider
+                    contextProvider.InitContextArgs(query.ContextArguments);
+                    httpContext.Response.ContentType = "text/event-stream";
```
Comment (dluc) [inline service/Service.AspNetCore/WebAPIEndpoints.cs:266]:
```suggestion
                    httpContext.Response.ContentType = "text/event-stream; charset=utf-8";
```

File: Dockerfile (hunk 4 above)
Comment (dluc) [reply Dockerfile]: fixed

File: service/Core/Core.csproj (hunk 6 above)
Comment (dluc) [reply service/Core/Core.csproj]: fixed

File: Dockerfile (hunk 5 above)
Comment (dluc) [reply Dockerfile]: fixed

File: service/Service/Program.cs (hunk 7 above)
Comment (dluc) [reply service/Service/Program.cs]: fixed

File: service/Service/Program.cs (hunk 9 above)
Comment (dluc) [reply service/Service/Program.cs]: fixed

File: Dockerfile (hunk 8 above)
Comment (dluc) [reply Dockerfile]: fixed

File: Directory.Packages.props (hunk 3 above)
Comment (dluc) [reply Directory.Packages.props]: removed

**Comment by dluc** [comment]
I'll make some changes to see if we can merge it:
* move the [DONE] token at the end of the stream, separate from the stream chunks
* merge stream behavior into the existing `/ask` endpoint, using an optional boolean flag to choose whether to stream or not
* remove code duplication in SearchClient
* revisit how content moderation is implemented

File: service/Core/Search/SearchClient.cs [hunk 21]
```diff
@@ -174,6 +176,125 @@ public async Task<MemoryAnswer> AskAsync(
...
+            this._log.LogWarning("No memories available");
+            result.AskResult.NoResultReason = "No memories available";
+            yield return result.AskResult;
+            yield break;
+        }
+
+        this._log.LogTrace("{Count} records processed", result.RecordCount);
+        var charsGenerated = 0;
+        var wholeText = new StringBuilder();
+        await foreach (var x in this._answerGenerator.GenerateAnswerTokensAsync(question, result.Facts.ToString(), context, cancellationToken).ConfigureAwait(true))
```
[the remark this answers is not in the document]
Comment (carlodek) [reply service/Core/Search/SearchClient.cs]: I've changed x variable name into token

File: service/Service/Service.csproj [hunk 22]
```diff
@@ -53,4 +53,4 @@
         </Content>
     </ItemGroup>
 
-</Project>
```
[the remark this answers is not in the document]
Comment (carlodek) [reply service/Service/Service.csproj:56]: I've removed the unnecessary blank line

File: service/Abstractions/HTTP/SSE.cs [hunk 23]
```diff
@@ -0,0 +1,67 @@
...
+using System.IO;
+using System.Linq;
+using System.Runtime.CompilerServices;
+using System.Text;
+using System.Text.Json;
+using System.Threading;
+
+namespace Microsoft.KernelMemory.HTTP;
+
+public static class SSE
```
Comment (dluc) [inline service/Abstractions/HTTP/SSE.cs:14]:
```suggestion
// See https://developer.mozilla.org/docs/Web/API/Server-sent_events/Using_server-sent_events
public static class SSE
```

**Review by dluc — APPROVED** [review]

**Comment by dluc** [comment]
**Example request:**

```
curl 'http://127.0.0.1:9001/ask' --header 'Content-Type: application/json' \
    --data '{"question": "which storage engines can I use with Kernel Memory?", "stream": true }'
```

**Response:**
```
data: {"streamState":"append","question":"which storage engines can I use with Kernel Memory?","noResult":false,"text":"","relevantSources":[... cut ...]}

data: {"streamState":"append","noResult":false,"text":"The"}

data: {"streamState":"append","noResult":false,"text":" storage"}

data: {"streamState":"append","noResult":false,"text":" engines"}

data: {"streamState":"append","noResult":false,"text":" that"}

[...]

data: {"streamState":"append","noResult":false,"text":"work"}

data: {"streamState":"append","noResult":false,"text":" in"}

data: {"streamState":"append","noResult":false,"text":" progress"}

data: {"streamState":"append","noResult":false,"text":")"}

data: [DONE]
```
